pkg/proxy/topology.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// FilterTopologyEndpoint returns the appropriate endpoints based on the cluster
+// topology.
+// It uses currentNodeName in nodeLabels map to get the current node's labels
+// which contains topology information, and use topologyKeys to filter out the
+// appropriate endpoints.
+// If both target endpoint's node and current node has a label which key equals to
+// topologyKeys[0], and value equals to each other, the endpoint will be chosen;
+// Only if no endpoints been chosen for topologyKeys[0] should endpoints for
+// toplogyKeys[1] be considered, and so on.
+// If topologyKeys is specified, but all entries have no endpoints that match the
+// the topology of the current node, the service has no endpoints for the current
+// node, and connections should fails.
+// The special value "" in topologyKeys may be used to mean "any node", this
+// catch-all value, if used, only makes sense as the last value in the list.
+// If topologyKeys is not specified or empty, no topology constraints will be
+// applied, it will return all endpoints.
+func FilterTopologyEndpoint(currentNodeName types.NodeName, nodeLabels map[types.NodeName]map[string]string, topologyKeys []string, endpoints []Endpoint) []Endpoint {
+	if len(topologyKeys) == 0 {
+		return endpoints
+	}
+	currentLabels, found := nodeLabels[currentNodeName]
+	if !found {
+		return endpoints
+	}
+
+	filteredEndpoint := []Endpoint{}
+	for _, key := range topologyKeys {
+		if key == v1.TopologyKeyAny {
+			return endpoints
+		}
+		topologyValue, found := currentLabels[key]
+		if !found {
+			continue
+		}
+
+		for _, ep := range endpoints {
+			nodeName := ep.GetNodeName()
+			if nodeName == "" {
+				continue
+			}
+			targetLabels, found := nodeLabels[nodeName]
+			if !found {
+				continue
+			}
+			if value, found := targetLabels[key]; found && value == topologyValue {
+				filteredEndpoint = append(filteredEndpoint, ep)
+			}
+		}
+		if len(filteredEndpoint) > 0 {
```

Comment:
I personally think this reads better if you put this check at the beginning of the topologyKeys for loop.

Reply:
or maybe just 

```
if len(filteredEndpoint) > 0 {
    return filteredEndpoint
}
```

Reply:
Let's just return `filteredEndpoint` when it's not empty.